activerecord/test/cases/connection_pool_test.rb

```diff
@@ -1477,6 +1477,24 @@ def test_inspect_does_not_show_secrets
         pool&.disconnect!
       end
 
+      def test_checkout_callback_error_does_not_affect_permanent_lease_and_active_connection_state
+        msg = "some error"
+        proc_to_raise = -> { raise msg }
+        ActiveRecord::ConnectionAdapters::AbstractAdapter.set_callback(:checkout, :after, proc_to_raise)
+
+        assert_predicate @pool, :permanent_lease?
+        assert_not_predicate @pool, :active_connection?
+        begin
+          @pool.lease_connection
+        rescue Exception => e
+          assert_equal msg, e.message
+        end
```

Comment:
```suggestion
        begin
          @pool.lease_connection
        rescue Exception => e
          assert_equal msg, e.message
        end
```

This should use `assert_raise`.